### PR TITLE
Toggle Updates - React

### LINKED ISSUE #3295: Toggle Upgrades - Part 1 (React)
### Part 1

# React
- Change isDefaultOpen to be isOpen and allow the consumer to be able to have the toggle open by default and also close/open it whenever.
- Update prop names for `titleAddClasses`  and  `iconAddClasses`. Follow the convention of elementAdditionalClasses. Change title to triggerText.
- Button element needs type=button
- Fix eslint errors
- add handler function prop to run when toggle is clicked 
- change `title` prop to be `triggerText` since title causes a11y issues.
- add ability to add classes to the content

### Angular -  Part 2 will be complete in Issue: #3296

### PR DESCRIPTION
**Pull request by guern1kn**
## What does this PR do?
- [x] Change isDefaultOpen to be isOpen and allow the consumer to be able to have the toggle open by default and also close/open it whenever.
- [x] Update prop name for titleAddClasses to be titleAdditionalClasses
- [x] Update iconAddClasses to be iconAdditionalClasses
- [x] Button element needs type=button
- [x] Fix eslint errors
- [x] add handler function prop to run when toggle is clicked
- [x] change title prop to be triggerText since title causes a11y issues.
- [x] add ability to add classes to the content 
- [x] Create issue for deprecation - isDefaultOpen, titleAddClasses, iconAddClasses, title - https://github.com/sparkdesignsystem/Backlog/issues/1296

### Associated Issue
Fixes #3295 

## Please check off completed items as you work.
If a checklist item or section does not apply to your PR then please remove it.

### Documentation
 - [x] Update Spark Docs

### Code
 - [x] Build Component in React
 - [x] Unit Testing in React with `npm run test` in `react/` (100% coverage, 100% passing)

### Browser Testing (current version and 1 prior)
  - [x] Google Chrome
  - [x] Google Chrome (Mobile)
  - [x] Mozilla Firefox
  - [x] Mozilla Firefox (Mobile)
  - [x] Microsoft Internet Explorer 11 (only this specific version of IE)
  - [x] Microsoft Edge
  - [x] Apple Safari
  - [ ] Apple Safari (Mobile)
  
### Accessibility Testing
  - [ ] Axe browser extension
  - [ ] Jaws Inspect
  - [ ] VoiceOver (iOS) or Talkback (Android)

### Deploy Preview Reviewed and Approved
 - [ ] Design Team
 - [ ] Accessibility Team

### Screenshots
Add screenshots to help explain your PR if you'd like. However, this is not
expected.

### CODE REVIEW

File: react/src/components/toggle/SprkToggle.js [hunk 1]
```diff
@@ -9,10 +9,15 @@ import 'focus-visible';
...
     const { isDefaultOpen } = this.props;
+    let { isOpen } = this.props;
+    if (isDefaultOpen !== undefined && isOpen === undefined) {
+      isOpen = isDefaultOpen;
+    }
     this.state = {
-      isOpen: isDefaultOpen || false,
-      height: isDefaultOpen ? 'auto' : 0,
+      isOpen: isOpen || false,
+      height: isOpen ? 'auto' : 0,
```
Comment (RCopeland) [inline react/src/components/toggle/SprkToggle.js:20]: i would assign the new isOpen from the old isDefaultOpen, then you dont have to have logic to check both:
```suggestion
    const {
      isDefaultOpen,
      isOpen = isDefaultOpen,
    } = this.props;

    this.state = {
      isOpen: isOpen || false,
      height: isOpen ? 'auto' : 0,
```

File: react/src/components/toggle/SprkToggle.js [hunk 2]
```diff
@@ -45,34 +56,45 @@ class SprkToggle extends Component {
 
     const containerClasses = classnames('sprk-c-Toggle', additionalClasses);
 
+    // TODO: Remove titleAddClasses in issue #1296
     const titleClasses = classnames(
       'sprk-c-Toggle__trigger sprk-b-TypeBodyThree sprk-u-TextCrop--none',
-      titleAddClasses,
+      titleAdditionalClasses || titleAddClasses,
```
Comment (RCopeland) [inline react/src/components/toggle/SprkToggle.js:62]: dont need this check if you make the change above

File: react/src/components/toggle/SprkToggle.js [hunk 3]
```diff
@@ -45,34 +56,45 @@ class SprkToggle extends Component {
...
     );
-
+    // TODO: Remove iconAddClasses in issue #1296
     const iconClasses = classnames(
       'sprk-c-Icon--xl sprk-c-Icon--toggle sprk-u-mrs',
       { 'sprk-c-Icon--open': isOpen },
-      iconAddClasses,
+      iconAdditionalClasses || iconAddClasses,
```
Comment (RCopeland) [inline react/src/components/toggle/SprkToggle.js:68]: here too

File: react/src/components/toggle/SprkToggle.js [hunk 4]
```diff
@@ -45,34 +56,45 @@ class SprkToggle extends Component {
...
+    );
     return (
       <div data-id={idString} {...other} className={containerClasses}>
         <button
           className={titleClasses}
           data-analytics={analyticsString}
-          onClick={this.toggleOpen}
+          onClick={(e) => {
+            if (onClick) {
+              onClick();
```
Comment (RCopeland) [inline react/src/components/toggle/SprkToggle.js:82]:
```suggestion
              onClick(e);
```

File: react/src/components/toggle/SprkToggle.js [hunk 5]
```diff
@@ -45,34 +56,45 @@ class SprkToggle extends Component {
...
+            this.toggleOpen(e);
+          }}
           aria-expanded={isOpen ? 'true' : 'false'}
           aria-controls={uniqueIdentifier}
           type="button"
         >
           <SprkIcon iconName={toggleIconName} additionalClasses={iconClasses} />
-          {title}
+          {/* TODO: Remove title in issue #1296 */}
+          {triggerText || title}
```
Comment (RCopeland) [inline react/src/components/toggle/SprkToggle.js:92]: dont need the check if you make the change above

**Review by RCopeland — CHANGES_REQUESTED** [review]
made a couple suggestions

**Comment by RCopeland** [comment]
@guern1kn I also see the aria-controls and matching ids are incrementing when ever its toggled open. They still match, but its weird. I think its because the unique id generation happens in the render, you should move it up to the constructor probably and make it part of state instead.

File: react/src/components/toggle/SprkToggle.js [hunk 6]
```diff
@@ -92,34 +108,61 @@ SprkToggle.propTypes = {
...
-  /** Determines if the toggle is open upon loading on the page. */
+  /**
+   * Deprecated - Use `isOpen` instead.
+   * Determines if the toggle is open upon loading on the page.
+   */
+  // TODO: isDefaultOpen in issue #1296
   isDefaultOpen: PropTypes.bool,
-   /**
-   * A space-separated string of classes to add to the outermost container of the component.
+  /** Determines if the toggle is open upon loading on the page. */
```
Comment (afebbraro) [inline react/src/components/toggle/SprkToggle.js:138]: hey @guern1kn , part of the a/c is that they can have it be open or closed whenever.  Lets update this prop description so it mentions that its not just on page load.

**Review by afebbraro — CHANGES_REQUESTED** [review]

**Comment by afebbraro** [comment]
@guern1kn the sprkaward component uses sprktoggle. Lets update that one line of 'title' in there too.

File: react/src/components/toggle/SprkToggle.js [hunk 7]
```diff
@@ -9,10 +9,12 @@ import 'focus-visible';
...
   constructor(props) {
     super(props);
-    const { isDefaultOpen } = this.props;
+    // TODO: Remove isDefaultOpen in issue #1296
+    const { isDefaultOpen, isOpen = isDefaultOpen } = this.props;
+
     this.state = {
-      isOpen: isDefaultOpen || false,
-      height: isDefaultOpen ? 'auto' : 0,
+      isOpen: isOpen || false,
```
Comment (bobwise) [inline react/src/components/toggle/SprkToggle.js:16]: Copying the prop into state in the constructor means any changes to the prop after init will be ignored.

**Review by bobwise — CHANGES_REQUESTED** [review]

**Review by afebbraro — APPROVED** [review]
Approved! Just need that update to the addclasses prop description then good for merge. Thanks.

**Review by RCopeland — APPROVED** [review]

**Pull request merged**